### DISCUSSION TITLE
PyStein Bug Bash Discussion

### DISCUSSION THREAD

**Discussion by vrdmr** [Azure/azure-functions-python-worker#975] [General]
A general discussion page created to capture any bugs reported in BugBash for PyStein

**Comment by stefanushinardi** [comment]
## Error handling

When my ```function_app.py``` has errors in it, instead of getting error messages, I just got warnings that said there is 0 function loaded. 


### Repro steps
1. Have a working app 
2. Add illegal syntax
3. run func host start
4. look at the console output

**Reply by surgupta-msft** [reply]
In my case, it showed me the errors. Not sure if our scenario is exactly the same.

**Scenario description** - function_app.py has 4 functions. For one of the functions, a required library (datetime) is not imported.

**Output** - All functions were loaded, but during the execution of timer trigger it gave exception.

After I imported the required library, it worked fine. Let me know if I am missing something.

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/16623184/158678735-6ffb61f4-f6ba-4851-b9d2-779911c4bc37.png">

**Reply by stefanushinardi** [reply]
interesting. Maybe my log level isnt setup properly, let me try that

**Reply by surgupta-msft** [reply]
I found another scenario which matches to the one you mentioned - 

**Scenario description** - route is set to var and var is not declared/assigned anywhere.

**Output** - Got message "0 job functions found" and did not get the actual error like - var is not defined.

```
@app.function_name(name="HttpTrigger2")
@app.route(route=var) # HTTP Trigger
def test_function2(req: func.HttpRequest) -> func.HttpResponse:
     return func.HttpResponse("HttpTrigger2 function processed a request!!!")
```

<img width="1237" alt="image" src="https://user-images.githubusercontent.com/16623184/158690368-852c7785-f84f-4095-a158-17631dafb762.png">

**Comment by stefanushinardi** [comment]
## Flask wsgi adpator error

Got the following error:

```
(.venv) C:\Users\shinardi\Downloads\python-functions-new-prg-model-main\python-functions-new-prg-model-main>C:\Users\shinardi\Downloads\CoreTools-PyStein-win\CoreTools-New-Prg-Model\func host start
Found Python version 3.7.9 (py).

Azure Functions Core Tools
Core Tools Version:       4.0.4395 Commit hash: N/A Integration build number: PreRelease220311-1913 (64-bit)   
Function Runtime Version: 4.3.0.17895

[2022-03-16T20:00:28.291Z] Worker process started and initialized.
[2022-03-16T20:00:28.448Z] A host error has occurred during startup operation '30fa724c-efd0-4412-91b4-a0e03909b239'.
[2022-03-16T20:00:28.452Z] Microsoft.AspNetCore.Routing: An error occurred while creating the route with name 'http_app_func' and template 'api//{*route}'. Microsoft.AspNetCore.Routing: The route template separator character '/' cannot appear consecutively. It must be separated by either a parameter or a literal value. (Parameter 'routeTemplate'). Microsoft.AspNetCore.Routing: The route template separator character '/' cannot appear consecutively. It must be separated by either a parameter or a literal value.
[2022-03-16T20:00:28.475Z] Failed to stop host instance 'a380bbfc-9b9d-4cf9-a404-d8e752f4401d'.
Value cannot be null. (Parameter 'provider')
[2022-03-16T20:00:28.476Z] Microsoft.Azure.WebJobs.Host: The host has not yet started.
```

Code:

```python
from flask import Flask, make_response, request
flask_app = Flask(__name__)
@flask_app.route("/")
def index():
    return (
        "Try /hello/Chris for parameterized Flask route.\n"
        "Try /module for module import guidance"
    )

@flask_app.route("/hello/<name>", methods=['GET','POST','DELETE'])
def hello(name: str):
    return f"hello {name}"
app = func.FunctionApp(wsgi_app=flask_app.wsgi_app, auth_level=func.AuthLevel.ANONYMOUS)
```

**Reply by stefanushinardi** [reply]
Issue resolved by overriding host.json settings

```json
"extensions" : {
    "http": {
      "routePrefix": ""
    }
  }
```

**Comment by surgupta-msft** [comment]
## Incorrect HttpTrigger URL

If route is empty (or only spaces), the Httptrigger URL is incorrect

Trigger code - 

```
@app.function_name(name="HttpTrigger2")
@app.route(route="") # HTTP Trigger
def test_function2(req: func.HttpRequest) -> func.HttpResponse:
     return func.HttpResponse("HttpTrigger2 function processed a request!!!")
```

<img width="555" alt="image" src="https://user-images.githubusercontent.com/16623184/158681877-8eef1037-5ac6-4f1b-953c-1e4376b1c5b8.png">


Other cases - 

1. if route = "//" then URL -> http://localhost:7071/api/\ -> Failed when browsed
2. if route = "\_." then URL -> http://localhost:7071/api/_. -> Worked when browsed

<img width="299" alt="image" src="https://user-images.githubusercontent.com/16623184/158691837-76c2c1cf-faa2-4639-bb6d-ab3be4076810.png">


<img width="322" alt="image" src="https://user-images.githubusercontent.com/16623184/158691884-07cad4cd-85db-4eec-b544-93def49443df.png">

**Comment by TsuyoshiUshio** [comment]
Without AzureWebJobsStrage, with Extension Bundle and http trigger (not using durable functions) we see the following issue. 

```
[2022-03-16T20:13:19.752Z] A host error has occurred during startup operation '60334426-b31c-4f9f-bad4-0f6af839cb9b'.
[2022-03-16T20:13:19.757Z] Microsoft.Azure.WebJobs.Extensions.DurableTask: Unable to find an Azure Storage connection string to use for this binding.
Value cannot be null. (Parameter 'provider')
```

**Comment by vrdmr** [comment]
- [Function File nameing](https://microsoft-my.sharepoint.com/:w:/p/vameru/EXcVUDxPjn9Pu-1WWCHppbcBW_QrppTx7jjf0Zuy1zTZOg?e=ygDU8D&nav=eyJjIjoxNTYzMzM2NTI1fQ)
